### DISCUSSION TITLE
How to Add Radio Stations?

### DISCUSSION THREAD

**Discussion by d4yz** [music-assistant/support#633] [Q&A] [answered]
How do you add Web Radio stations to the Radio section?

**Comment by jazzyisj** [comment, marked as answer]
- Create a free account on Tune In and add some stations to your favorites on the Tune In website.
- Set tune in up with your login credentials in Music Assistant.
- Wait for a bit for the database to sync.
- Your favourite stations on Tune In should appear in Music Assistant

https://tunein.com/

**Reply by d4yz** [reply]
ahh so this is just for Tune-in Radio? 

Is there a way to add custom internet radio stations to this list?

**Reply by jazzyisj** [reply]
There is now way to add custom stations yet.  The only providers available are the ones shown in the integration settings.
If you'd like to see this feature, I'd create a new Feature Request discussion so people can vote on it!

**Reply by marcelveldt** [reply]
1) You can also add custom radio station url's to Tunein if you want.

2) In fact custom url stations are supported internally but there is not yet any UI exposed to add the URL's. There is however a little trick you can try. Just create a m3u playlist with your url's in there. Just the URL's are enough. Add that m3u file (name it to your liking) in the directory you have added to MA for the "local files" music providers and voila, you should see your radio stations.

**Reply by jazzyisj** [reply]
> add custom radio station url's to Tunein

Just an FYI - this can't be done on Windows, you have to use the anrdroid or ios app.

https://help.tunein.com/where-can-i-add-a-custom-url-SJ1Gg5C5_Df

M3U trick works great!

**Reply by DutchJaFO** [reply]
Will there be a way to integrate the internet radio stations from the HA radio browser integration ?
Or is Tunein the way to go (for now?)?
Must admit I hadn't thought about the m3u file trick..

**Reply by marcelveldt** [reply]
Yes, the endgoal is that you can add radio stations from HA radio browser to MA favorites. 

In fact now that I think of it that might also already work since yesterdays 2022.7 beta ..

Simply play a station from HA radio browser to the MA player.
I guess that after that you can find that radio station listed in the radio section of MA (deselect the favorites filter) and you can press the heart to favorite it. Untested!

**Reply by jazzyisj** [reply]
>Simply play a station from HA radio browser to the MA player.

Works!  The first time I tried it took a long time to start and the following was in the log.  Playing subsequent stations there was almost no delay and no errors in the log.
```
2022-07-13 18:25:31 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [281471509916176] coroutine raised StopIteration
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/music_assistant/helpers/tags.py", line 206, in parse_tags
    return AudioTags.parse(data)
  File "/usr/local/lib/python3.10/site-packages/music_assistant/helpers/tags.py", line 147, in parse
    audio_stream = next(x for x in raw["streams"] if x["codec_type"] == "audio")
StopIteration
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 193, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1713, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1750, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 680, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 930, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 717, in _handle_entity_call
    await result
  File "/config/custom_components/mass/media_player.py", line 427, in async_play_media
    await self.player.active_queue.play_media(media_id, queue_opt)
  File "/usr/local/lib/python3.10/site-packages/music_assistant/models/player_queue.py", line 219, in play_media
    media_item = await self.mass.music.get_item_by_uri(item)
  File "/usr/local/lib/python3.10/site-packages/music_assistant/controllers/music/__init__.py", line 216, in get_item_by_uri
    return await self.get_item(
  File "/usr/local/lib/python3.10/site-packages/music_assistant/controllers/music/__init__.py", line 237, in get_item
    return await self.get_provider(ProviderType.URL).parse_item(item_id)
  File "/usr/local/lib/python3.10/site-packages/music_assistant/music_providers/url.py", line 92, in parse_item
    item_id, url, media_info = await self._get_media_info(
  File "/usr/local/lib/python3.10/site-packages/music_assistant/music_providers/url.py", line 143, in _get_media_info
    media_info = await parse_tags(url)
RuntimeError: coroutine raised StopIteration
```
> find that radio station listed in the radio section of MA 

Doesn't work.  Played radio stations from radio browser do not appear in MA radio section.  Tried Run Sync, disabled fav filter, waited, restarted.

**Reply by marcelveldt** [reply]
OK, in that first case I think that was a radio station providing an m3u file (playlist) instead of an audio stream. I already have an issue report open for that I need to address. Can you tell me which radio station you tried ?

As for the second... it was a wild guess from me if that actually already worked. I now see that I actively block that from happening but I can put a small "like button" (heart icon) in the OSD (playbar below)

**Reply by jazzyisj** [reply]
>Can you tell me which radio station you tried ?

Shoot, I can't off the top of my head.  I was just randomly clicking stations.  I will spend some time today once the house is awake to see if I can recreate the error again for you.

**Reply by jazzyisj** [reply]
Tried dozens of stations and I was not able to reproduce this error particular again.  Assume you got it fixed in the betas somewhere.

**Comment by Symbiot78** [comment]
some findings regarding radiostations and MA..

1. I added a m3u file to my local music library (wish I could add multiple file system providers)
2. ran a sync
3. radiostations(filesystem) showed up
4. I can browse it and start station

I wanted to use this in an automation, so I created a test automation
used 'play media'
selected a mass player
clicked 'pick media' -> music assistant -> playlists
no radio station playlist was visible.
compared the visible with my playlists..
Apparently pick media->mass -> playlists only shows playlists that are favorited in MASS.
added my radio stations m3u as a favorite.
it  now shows up
Clicked the 'radiostations'  | unknown error
then clicked the little '+' 
that worked

but of course.. if the file contains more than 1 url you're screwed.

**Reply by OzGav** [reply]
Looks like you have solved this yourself and thanks for the details which may help others. I don't think it is a problem having a playlist with one url. If you would like a better way first have a look at the various feature requests to see if someone else has suggested it and if not put in your own request. 👍🏻

**Reply by Symbiot78** [reply]
thanks .. I'll try. 
I see creating 1 m3u pr. radiostation as an ok solution.. it's a 1 time thing and I don't imagine wanting 100 radiostations.
Of course having a more fluid way of adding radiostations would be perfect but I can understand it's not something you just implement in 1 hour.

**Reply by OzGav** [reply]
If the stations are available in TuneIn you can favourite them all there and they will then appear in the RADIO tab

**Reply by Symbiot78** [reply]
I know but that adds commercial blocks to an otherwise commercial free station .. so to speaks

**Reply by OzGav** [reply]
Fair enough. The url in a Playlist appears to be a solution then?

**Reply by OzGav** [reply]
There is a feature request for TuneIn premium to get rid of the ads by the way https://github.com/music-assistant/hass-music-assistant/discussions/947

**Reply by Symbiot78** [reply]
@OzGav  -  the url works. However I've learned that not all url's are equal.. or at least the data/meta provided by the urls are not all very usable.
I added BBC radio 2.. that shows up perfectly.. and even under "radio stations" in MA. 
Then I added a local national radio station.. That comes up as "stream900".. or similar.. if I add more from the local radio network - they come up as the same.
I haven't found a way to tag them with a proper name. I added all sorts of info in the M3u file but to no avail.

**Reply by OzGav** [reply]
When you say you have added “all sorts of info” do you mean extended m3u tags? https://windowsreport.com/m3u-playlist-guide/
https://docs.fileformat.com/audio/m3u/

**Comment by Symbiot78** [comment]
Sorry.. yeah that's it.

for instance:

`#EXTM3U

#EXTINF: -1, DR P8 Jazz
https://drliveradio.akamaized.net/hls/live/2022411/p8jazz/masterab.m3u8

#EXTINF: 0, DR P4 Nordjylland
https://drliveradio.akamaized.net/hls/live/2022411/p4nordjylland/masterab.m3u8`

**Reply by OzGav** [reply]
Ok will have to see whether @marcelveldt has added support for those extra tags.

**Comment by oneofthemany** [comment]
If you install the Radio Browser integration and play a channel it will show up in the radio player within Music Assistant.

You can then simply just favourite the channel and Music Assistant holds a persistent config based on the newly created favourite channel.

An alternative is to create a script and in the sequence section use the `service: media_player.play_media`, with the target as MASS, add the URL to the content ID with music as the content type:

<img width="1139" alt="image" src="https://user-images.githubusercontent.com/44924639/200581099-041d3d2b-810a-41a2-925b-a90d40fe6c19.png">

Run the script, go to MASS, then simply add to favourites within the Radio section

**Comment by OzGav** [comment]
Old v1 info
